src/ray/gcs/gcs_server/gcs_task_manager.cc

```diff
@@ -27,6 +27,37 @@ void GcsTaskManager::Stop() {
   }
 }
 
+std::vector<rpc::TaskEvents> GcsTaskManager::GcsTaskManagerStorage::GetTaskEvents(
+    const absl::flat_hash_set<std::string> &task_ids,
+    absl::optional<std::string> job_id) {
+  std::vector<rpc::TaskEvents> results;
+
+  if (!task_ids.empty()) {
+    RAY_CHECK(!job_id.has_value())
+        << "only one of job_id or task_ids should be provided. ";
+    // Get all task events with task ids.
+    for (const auto &task_event : task_events_) {
+      if (task_ids.contains(task_event.task_id())) {
+        results.push_back(task_event);
+      }
+    }
+    return results;
+  }
+
+  if (job_id) {
+    // Get all task events with job id
+    for (const auto &task_event : task_events_) {
+      if (task_event.has_task_info() && task_event.task_info().job_id() == *job_id) {
```

Comment:
Same question here. We should have an index by job_id. How about changing our data structure to store tasks as a `Map[JobID -> Map[TaskID -> TaskData]]`? We can require queries by task_id to include a job id.

Reply:
Right - I should have doc it. 

I was thinking if we want to measure the performance e2e first (e.g. if retrieving with O(n)) is too slow to serve the dashboard. 

The major concern I had with adding a job->id index is this will add overheads to write path, which probably needs to be more scalable than read? (Since we might have hundreds of workers pushing data to GCS concurrently at an inverval of 1sec default).

Reply:
I don't think the index should be adding appreciable overheads--- a hash table lookup is probably at most 50ns of overhead. Typically you'd use a LRUMap type of data structure here, such as `LRUMap[JobID -> LRUMap[TaskID, TaskData]]`.

Reply:
(And the JobID->tasks index is likely to be in CPU cache and hence even lower overhead.)

Reply:
Thinking about it again - I guess the overhead with what you suggested is probably minimal. (I was thinking of having 2 separate index)

This also yields the benefits that we could enforce the number of task events sotred per job easily here.